src/UbiMesh.cpp

```diff
@@ -0,0 +1,320 @@
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+Developed and maintained by Jose Garcia for Ubidots Inc
+@jotathebest at github: https://github.com/jotathebest
+*/
+
+#include "UbiMesh.h"
+#include "UbiProtocolHandler.h"
+
+/**************************************************************************
+ * Static Functions
+ ***************************************************************************/
+#if PLATFORM_ID != PLATFORM_XENON && PLATFORM_ID != PLATFORM_XENON_SOM && \
+    PLATFORM_ID != PLATFORM_PHOTON_DEV &&                                 \
+    PLATFORM_ID != PLATFORM_PHOTON_PRODUCTION &&                          \
+    PLATFORM_ID != PLATFORM_ELECTRON_PRODUCTION
+bool UbiMesh::isThrottled() {
+  // Simple throttling routine for TCP and HTTP
+  bool throttled = true;
+  if (iotProtocolMesh == UBI_HTTP || iotProtocolMesh == UBI_TCP) {
+    if (millis() - time_now > 20000) {
+      throttled = false;
+      time_now = millis();
+    }
+  }
+
+  if (iotProtocolMesh == UBI_UDP || iotProtocolMesh == UBI_PARTICLE) {
+    throttled = false;
+  }
+
+  return throttled;
+}
+
+void UbiMesh::ubiPublishHandler(const char* event, const char* data) {
+  bool throttled = isThrottled();
+
+  if (throttled) {
+    if (_debugMesh) {
+      Serial.println(
+          "[WARNING] your sample time reached the throttling of 20 s, please "
+          "set a higher sample time or use UDP or Particle Webhooks to send "
+          "data");
+    }
+  }
+
+  if (!throttled) {
+    uint8_t i = 0;
+    char* _data = const_cast<char*>(data);
+    char* pch;
+    const char _meshDelimiter[2] = "|";
+    pch = strtok(_data, _meshDelimiter);
+
+    std::map<uint8_t, char*> meshMap;
+
+    while (pch != NULL) {
+      meshMap.insert(std::pair<uint8_t, char*>(i, pch));
+      i++;
+      pch = strtok(NULL, _meshDelimiter);
+    }
+
+    MeshUbi* dots = (MeshUbi*)malloc(sizeof(MeshUbi));
+    UbiMesh* _protocolInternalMesh = new UbiMesh(_tokenMesh);
+    UbiProtocolHandler* _meshCloudHandler =
+        new UbiProtocolHandler(_tokenMesh, iotProtocolMesh);
+    _protocolInternalMesh->buildDots(meshMap, dots);
+    _meshCloudHandler->setDebug(true);
+    _meshCloudHandler->add(dots->variableLabel, dots->dotValue,
+                           dots->dotContext, dots->dotTimestampSeconds,
+                           dots->dotTimestampMillis);
+    _meshCloudHandler->send(dots->deviceLabel, dots->deviceName);
+    delete _meshCloudHandler;
+    delete _protocolInternalMesh;
+    free(dots);
+  }
+}
+#endif
```

Comment:
Why close this if here? I think this should go to the end of the file

Reply:
The class and all its methods should be defined once